keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades/README.md

```diff
@@ -0,0 +1,844 @@
+more complex scenarios (eg. topology spread associating two deployments at 
+the same time), User need to be responsible for providing MatchedKeys to 
+identify which pods should be grouped. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+In addition to using `pod-template-hash` added by the Deployment controller, 
+users can also provide the customized key in  `MatchLabelKeys` to identify 
+which pods should be grouped. If so, the user needs to ensure that it is 
+correct and not duplicated with other unrelated workloads.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+`TopologySpreadConstraint` needs an extra field. The user fills in the 
+`MatchLabelKeys` with the key of label. The scheduler plugin will look up 
+the value from the incoming pod labels. Pods with the same key-value pairs 
+will be treated as groups. The pods belonging to the same group will be part 
+of the spreading in PodTopologySpread.
+
+A new field named `MatchLabelKeys` will be introduced to`TopologySpreadConstraint`:
+```go
+type TopologySpreadConstraint struct {
+	MaxSkew           int32
+	TopologyKey       string
+	WhenUnsatisfiable UnsatisfiableConstraintAction
+	LabelSelector     *metav1.LabelSelector
+
+	// MatchLabelKeys holds keys to labels whose values we did not know
+	// before creating the pod. The scheduler plugin will obtain this
+	// value from the incoming pod labels. Pods with the same key-value
+	// pairs are considered a group. In the PodTopologySpread, the
+	// spreading will include pods belonging to the same group. It is
+	// an array of strings, each element representing a key of the
+	// label. If the key does not exist, the element won't be ignored.
+	MatchLabelKeys []string
+}
+```
+
+Examples of use are as follows:
+```yaml
+     topologySpreadConstraints:
+        - maxSkew: 1
+          topologyKey: kubernetes.io/hostname
+          whenUnsatisfiable: DoNotSchedule
+          matchLabelKeys:
+            - app
+            - pod-template-hash
+```
+
+The scheduler plugin `PodTopologySpread` will obtain the labels from the pod 
+labels by the keys in `matchLabelKeys`. The obtained labels will be merged 
+to `labelSelector` of `topologySpreadConstraints` to filter and group pods. 
+The pods belonging to the same group will be part of the spreading in 
+`PodTopologySpread`.
+
```

Comment:
Please mention the following:

```
Finally, the feature will be guarded by a new feature flag. If the feature is disabled, the field MatchLabelKeys 
is preserved if it was already set in the persisted Pod object, otherwise it is silently dropped; moreover, 
kube-scheduler will ignore the field and continue to behave as before.
```